kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala

```diff
@@ -99,25 +100,34 @@ class BatchJobSubmission(session: KyuubiBatchSessionImpl, batchRequest: BatchReq
     } catch onError("submitting batch job submission operation in background, request rejected")
   }
 
+  private def applicationFailed(applicationStatus: Option[Map[String, String]]): Boolean = {
+    applicationStatus.map(_.get(ApplicationOperation.APP_STATE_KEY)).exists(s =>
+      s.contains("KILLED") || s.contains("FAILED"))
+  }
+
   private def submitBatchJob(): Unit = {
     try {
       info(s"Submitting ${batchRequest.batchType} batch job: $builder")
       val process = builder.start
       var applicationStatus = currentApplicationState
-      while (applicationStatus.isEmpty && process.isAlive) {
+      while (!applicationFailed(applicationStatus) && process.isAlive) {
+        process.waitFor(applicationCheckInterval, TimeUnit.MILLISECONDS)
         applicationStatus = currentApplicationState
-        Thread.sleep(applicationCheckInterval)
       }
 
-      if (applicationStatus.map(_.get(ApplicationOperation.APP_STATE_KEY)).filter(s =>
-          s == Some("KILLED") || s == Some("FAILED")).isDefined) {
+      if (applicationFailed(applicationStatus)) {
         process.destroyForcibly()
         throw new RuntimeException("Batch job failed:" + applicationStatus.get.mkString(","))
       } else {
-        process.waitFor()
         if (process.exitValue() != 0) {
           throw new KyuubiException(s"Process exit with value ${process.exitValue()}")
         }
+
+        applicationStatus = currentApplicationState
+        if (applicationFailed(applicationStatus)) {
```

Comment:
seems this branch is unnecessay ? we have checked status in if branch

Reply:
the while loop might terminate because the process is not alive.

So, the application status might not be latest.

So, It is better to get final application status at last.

Reply:
I think it's minor .. the check of `isActive` is a lightweight operation so the status should always be same

Reply:
ok